Sources/GRPCHTTP2Core/Server/Connection/ServerConnectionManagementHandler.swift

```diff
@@ -0,0 +1,473 @@
+
+    // Generate a random value to be used as keep alive ping data.
+    let pingData = UInt64.random(in: .min ... .max)
+    self.keepAlivePingData = HTTP2PingData(withInteger: pingData)
+
+    self.state = StateMachine(
+      allowKeepAliveWithoutCalls: allowKeepAliveWithoutCalls,
+      minPingReceiveIntervalWithoutCalls: minPingIntervalWithoutCalls,
+      goAwayPingData: HTTP2PingData(withInteger: ~pingData)
+    )
+
+    self.flushPending = false
+    self.inReadLoop = false
+    self.clock = clock
+    self.frameStats = FrameStats()
+  }
+
+  func handlerAdded(context: ChannelHandlerContext) {
+    assert(context.eventLoop === self.eventLoop)
+  }
+
+  func channelActive(context: ChannelHandlerContext) {
+    self.maxAgeTimer?.schedule(on: context.eventLoop) {
+      self.initiateGracefulShutdown(context: context)
+    }
+
+    self.maxIdleTimer?.schedule(on: context.eventLoop) {
+      self.initiateGracefulShutdown(context: context)
+    }
+
+    self.keepAliveTimer?.schedule(on: context.eventLoop) {
+      self.keepAliveTimerFired(context: context)
+    }
+
+    context.fireChannelActive()
+  }
+
+  func channelInactive(context: ChannelHandlerContext) {
+    self.maxIdleTimer?.cancel()
+    self.maxAgeTimer?.cancel()
+    self.maxGraceTimer?.cancel()
+    self.keepAliveTimer?.cancel()
+    self.keepAliveTimeoutTimer.cancel()
+    context.fireChannelInactive()
+  }
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    switch event {
+    case let event as NIOHTTP2StreamCreatedEvent:
+      // The connection isn't idle if a stream is open.
+      self.maxIdleTimer?.cancel()
+      self.state.streamOpened(event.streamID)
+
+    case let event as StreamClosedEvent:
+      switch self.state.streamClosed(event.streamID) {
+      case .startIdleTimer:
+        self.maxIdleTimer?.schedule(on: context.eventLoop) {
+          self.initiateGracefulShutdown(context: context)
+        }
+
+      case .close:
+        context.close(mode: .all, promise: nil)
+
+      case .none:
+        ()
+      }
+
+    default:
+      ()
+    }
+
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    self.inReadLoop = true
+
+    // Any read data indicates that the connection is alive so cancel the keep-alive timers.
+    self.keepAliveTimer?.cancel()
+    self.keepAliveTimeoutTimer.cancel()
+
+    let frame = self.unwrapInboundIn(data)
+    switch frame.payload {
+    case .ping(let data, let ack):
+      if ack {
+        self.handlePingAck(context: context, data: data)
+      } else {
+        self.handlePing(context: context, data: data)
+      }
+
+    default:
+      ()  // Only interested in PING frames, ignore the rest.
+    }
+
+    context.fireChannelRead(data)
+  }
+
+  func channelReadComplete(context: ChannelHandlerContext) {
+    while self.flushPending {
+      self.flushPending = false
+      context.flush()
```

Comment:
This is more of a NIO question but want to double check: `context.flush()` will start by calling `self.flush()` in this case, right? Which would in turn call `maybeFlush(context:)`, which can reset `flushPending` to be `true` again. 
I just want to make sure I understand why we're checking for `flushPending` in a `while` loop.

Reply:
No, `context.flush()` will call `flush(context:)` on the _next_ channel handler. 

We're calling it in a loop here in case a flush re-entrantly triggers another write which could set `flushPending` back to true.